agent/util-scripts/pbench-tool-meister

```diff
@@ -0,0 +1,998 @@
+    """
+
+    @staticmethod
+    def fetch_params(params):
+        """Static help method that allows the method constructing a ToolMeister
+        instance to verify the parameters before we actually construct the
+        object.
+
+        The definition of the contents of a parameter block is really
+        independent of a ToolMeister implementation, but we keep this method
+        in the ToolMeister class since it is closely related to the
+        implementation.
+        """
+        try:
+            benchmark_run_dir = params["benchmark_run_dir"]
+            channel = params["channel"]
+            controller = params["controller"]
+            group = params["group"]
+            hostname = params["hostname"]
+            tools = params["tools"]
+        except KeyError as exc:
+            raise ToolMeisterError(f"Invalid parameter block, missing key {exc}")
+        else:
+            return benchmark_run_dir, channel, controller, group, hostname, tools
+
+    def __init__(self, pbench_bin, params, redis_server, logger):
+        self.logger = logger
+        self.pbench_bin = pbench_bin
+        ret_val = self.fetch_params(params)
+        (
+            self._benchmark_run_dir,
+            self._channel,
+            self._controller,
+            self._group,
+            self._hostname,
+            self._tools,
+        ) = ret_val
+        self._running_tools = dict()
+        self._rs = redis_server
+        logger.debug("pubsub")
+        self._pubsub = self._rs.pubsub()
+        logger.debug("subscribe %s", self._channel)
+        self._pubsub.subscribe(self._channel)
+        logger.debug("listen")
+        self._chan = self._pubsub.listen()
+        logger.debug("done listening")
+        # Name of the temporary tool data directory to use when invoking
+        # tools.  This is a local temporary directory when the Tool Meister is
+        # remote from the pbench controller.
+        if self._controller != self._hostname:
+            self._tmp_dir = Path(
+                tempfile.mkdtemp(
+                    dir=os.environ.get("_PBENCH_TOOL_MEISTER_TMP", "/var/tmp"),
+                    prefix=f"tm.{self._group}.{os.getpid()}.",
+                )
+            )
+        else:
+            self._tmp_dir = None
+        # Now that we have subscribed to the channel as specified in the
+        # params object, we need to pull off the first message, which is an
+        # acknowledgement that we have properly subscribed.
+        logger.debug("next")
+        resp = next(self._chan)
+        assert resp["type"] == "subscribe", f"Unexpected 'type': {resp!r}"
+        assert resp["pattern"] is None, f"Unexpected 'pattern': {resp!r}"
+        assert (
+            resp["channel"].decode("utf-8") == self._channel
+        ), f"Unexpected 'channel': {resp!r}"
+        assert resp["data"] == 1, f"Unexpected 'data': {resp!r}"
+        logger.debug("next done")
+        # First state we are expecting
```

Comment:
Watching some of your walkthrough, I was wondering if we should be encapsulating some of the redis behavior in a class; it'd make callers easier to read by presenting a semantic abstraction rather than a half page of details, and might also serve to allow more consistency (and maintainability). 

This, for example, is a "validated subscribe" (which may be the only form of subscribe we need) that validates we receive the initial "subscribe" method and fails either if the subscription itself failed or if we don't get the confirmation.

```
self._redis = PbenchRedis(redis_server)
self._redis.subscribe() 
```
would be a lot cleaner to read. The class could also handle spawning and shutting down the redis server, etc.

OK, maybe a separate follow-up PR; but something to think about.

Reply:
I'd be happy to implement this abstraction, as I think it is a good idea.  Should we gate this PR on having this?  I might not be too bad to do, but just adding more time.

Reply:
Separate PR is fine by me.

Reply:
Tracked via issue #1781.